### PR TITLE
Kyra/win game

### PR DESCRIPTION
**Pull request by kyralee22**
This PR currently has the winning score threshold set as 4 points so you guys can see what it looks like when a player wins. It will be changed to 100 later. Also, I could not figure out how to restart the game so the commented out parts are my attempts (extracting the main function to another file/module made a dependent cycle, using a ref to house the main function to run later didn't work, I ran into a bunch of logic issues). If you guys can give me guidance on how to restart the game either using main or running the loop with a fresh game state that would be helpful :)

### CODE REVIEW

**Review by isabellawu90 — APPROVED** [review]

File: bin/main.ml [hunk 1]
```diff
@@ -449,6 +449,62 @@ let draw_players ctx state =
...
+        LTerm_style.none with
+        background = Some LTerm_style.white;
+        bold = Some true;
+      }
+    label
+(*LTerm_draw.draw_string ctx ((ctx_size.rows / 2) + 3) ((ctx_size.cols /
+  2) - 18) ~style:{ LTerm_style.none with background = Some
+  LTerm_style.red } (Zed_string.of_utf8 "PRESS ENTER TO REPLAY OR DELETE
+  TO EXIT");
+
```
Comment (owenxr) [inline bin/main.ml:487]: Can we test the replay code as well?

**Review by owenxr — CHANGES_REQUESTED** [review]

**Comment by owenxr** [comment]
For restarting the game after a win:

Instead of thinking about rerunning the main function, think about how you can reset the game state to mimick the beginning of a game. After the `draw` function is run, the `loop` function will be run afterwards. This means even on the win screen, all the keys are still able to be used. We can take advantage of this to return a new game state that mimicks the beginning of a game, essentially creating a restart.

Trying creating field in `game_state` within `state.ml` called `mutable in_play`. You will want to make it mutable because you will want to change this state when you are drawing the win_box. Then, for example, in the `loop` function you can go to the `Enter` key event match and insert a boolean to check this `in_play` state. If `in_play` implies a win condition is met, `LoopResultUpdate` the current state except resetting:
- board to a new board
- players to players with 0 points and a new deck
- entry = SelectStart
- current_player_index = 0
- dict (this should remain the same so nothing to change here)
- in_play reset to represent no winning condition met

**Comment by owenxr** [comment]
Screen Shot 2021-11-30 at 9 14 19 PM

**Comment by owenxr** [comment]
> This PR currently has the winning score threshold set as 4 points so you guys can see what it looks like when a player wins. It will be changed to 100 later. Also, I could not figure out how to restart the game so the commented out parts are my attempts (extracting the main function to another file/module made a dependent cycle, using a ref to house the main function to run later didn't work, I ran into a bunch of logic issues). If you guys can give me guidance on how to restart the game either using main or running the loop with a fresh game state that would be helpful :)

Game restart should be implemented with my commit

**Comment by laptou** [comment]
I think that the win condition should be checked in the update loop and not in the draw function. It's better to keep the state updating code and the drawing code as separate as possible to avoid confusion, and that would make this more consistent with how everything else works. Since there's only one place in the update loop where the player's score can increase (when they are finishing a word entry), you can just check if the new state is one where one of the players has won, and if so, then we return a state that indicates that the game is over.

We could also change `game_state` into an enum that has different variants for different parts of the game, ex.:

```ocaml
type game_state = 
  | Title { selection: title_screen_selection }
  | Gameplay { (* ... *) }
  | Victory { winning_player: int; scores: int list }
  | Options { (* ... *) }
```

**Review by laptou — CHANGES_REQUESTED** [review]
I think that the win condition should be checked in the update loop and not in the draw function. It's better to keep the state updating code and the drawing code as separate as possible to avoid confusion, and that would make this more consistent with how everything else works. Since there's only one place in the update loop where the player's score can increase (when they are finishing a word entry), you can just check if the new state is one where one of the players has won, and if so, then we return a state that indicates that the game is over.

We could also change `game_state` into an enum that has different variants for different parts of the game, ex.:

```ocaml
type game_state = 
  | Title { selection: title_screen_selection }
  | Gameplay { (* ... *) }
  | Victory { winning_player: int; scores: int list }
  | Options { (* ... *) }
```

**Comment by laptou** [comment]
superseded by #68

**Pull request closed without merging**